src/sqlparser/src/parser.rs

```diff
@@ -436,11 +436,28 @@ impl Parser {
                     UnaryOperator::Minus
                 };
                 let mut sub_expr = self.parse_subexpr(Self::PLUS_MINUS_PREC)?;
-                // TODO: Deal with nested unary exp: -(-(-(1))) => -1
-                // Tracked by: <https://github.com/singularity-data/risingwave/issues/4344>
+                while let Expr::Nested(expr1) = sub_expr {
+                    sub_expr = *expr1;
+                }
                 if let Expr::Value(Value::Number(ref mut s, _)) = sub_expr {
                     if tok == Token::Minus {
-                        *s = format!("-{}", s);
+                        // If we need to optimize for case of existing negative prefix,
+                        // we can opt to follow doNegateFloat here:
+                        // <https://github.com/postgres/postgres/blob/08909e3aee6182a988da5bf5c1b543910ee096c7/src/backend/parser/gram.y#L19128>
+                        // and simply increment the pointer, avoiding O(n) reallocation.
+                        // However this alternate approach might require unsafe rust
+                        // and some unstable features at the moment
+                        // (`#![feature(vec_into_raw_parts)]`).
+                        // Another alternative would be to refactor `Number`
+                        // to reserve a space for sign, however that means it can no longer
+                        // purely use a string as internal representation,
+                        // which adds complication downstream.
+                        let mut s_iter = s.chars();
+                        if s_iter.next().unwrap() == '-' {
+                            *s = s_iter.collect()
+                        } else {
+                            *s = format!("-{}", s);
```

Comment:
How about moving it to expression constant folding?

Reply:
Good suggestion.

After thinking about it more, here are some concerns:
When we reach constant folding pass, we operate on `ExprImpl`.
As such we expect number to have been inferred / cast to a concrete type. e.g. int, bigint, etc...

I have considered the following cases for this, which this PR aims to deal with:

| cases | postgres - pg_typeof result |
|-|-|
| `-9223372036854775808` | bigint |
| `-(-(-(9223372036854775808' | bigint |
| `-(+(9223372036854775808))` | bigint out of range |

For the first case, if we choose to handle it in constant folding pass. 
This means that `9223372036854775808` is first inferred as a numeric after parsing (as it is larger than max bigint).
Then we negate it into a negative numeric. This behaviour is different from postgres (postgres outputs bigint).

For the second case, the reasoning is the same as first case.

For final case, I just listed it as a note to indicate that postgres ignores leading `+` for integer types **during parsing**. It does remove it for floats, but so far that doesn't cause any issues for RW with regards to typing (in contrast to prefix `-`), so I'm leaving it to avoid further complications.

Reply:
FYI, PG still has `doNegate` in parser after it has constant folding:
```
/* doNegate()
 * Handle negation of a numeric constant.
 *
 * Formerly, we did this here because the optimizer couldn't cope with
 * indexquals that looked like "var = -4" --- it wants "var = const"
 * and a unary minus operator applied to a constant didn't qualify.
 * As of Postgres 7.0, that problem doesn't exist anymore because there
 * is a constant-subexpression simplifier in the optimizer.  However,
 * there's still a good reason for doing this here, which is that we can
 * postpone committing to a particular internal representation for simple
 * negative constants.	It's better to leave "-123.456" in string form
 * until we know what the desired type is.
 */
```

Personally I feel it good if `-9223372036854775808` can be assigned type correctly, but IMHO whether things like `-(-(-(9223372036854775808)))` should be handled the same as PG is not a must, since it's not a normal usage 😅🤪. e.g., in rust:
```
error: this arithmetic operation will overflow
 --> src/tmp.rs:7:23
  |
7 |         let a: i64 = -(-(-(9223372036854775808)));
  |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^ attempt to negate `i64::MIN`, which would overflow
  |
  = note: `#[deny(arithmetic_overflow)]` on by default

error: this arithmetic operation will overflow
 --> src/tmp.rs:8:23
  |
8 |         let b: i64 = ---9223372036854775808;
  |                       ^^^^^^^^^^^^^^^^^^^^^ attempt to negate `i64::MIN`, which would overflow

```

Reply:
>  In other words, it makes sense to me if it is parsed as neg neg i64::MIN and then rejected as out of range, instead of accepted as i64::MIN

Yes that makes more sense actually. I think PG behaviour seems unintuitive in this case vs rust.
In that case we don't even need this PR :rofl: , currently parser already ensures prefix negative maintain their prefix. so bigint::min is accepted.
If there's no other strong opinions, I will close this PR and #4344 .

Reply:
(Closing can re-open if there's further discussion to be made.)